File: python/bls_benchmark.py
#!/usr/bin/env python3
"""
BLS Signature Verification Benchmark for Ethereum Attestations

This script benchmarks how many BLS signatures of Ethereum attestations
can be verified per second on the current machine.
"""

import time
import random
import statistics
from typing import List, Tuple
from dataclasses import dataclass

from py_ecc.bls import G2ProofOfPossession as bls
from py_ecc.bls.g2_primitives import G1_to_pubkey, G2_to_signature
from py_ecc.optimized_bls12_381 import G1, G2, Z1, Z2

# Simplified Ethereum attestation data structure
@dataclass
class AttestationData:
    slot: int
    index: int
    beacon_block_root: bytes
    source_epoch: int
    source_root: bytes
    target_epoch: int
    target_root: bytes

    def serialize(self) -> bytes:
        """Serialize the attestation data to bytes for signing"""
        return b''.join([
            self.slot.to_bytes(8, 'little'),
            self.index.to_bytes(8, 'little'),
            self.beacon_block_root,
            self.source_epoch.to_bytes(8, 'little'),
            self.source_root,
            self.target_epoch.to_bytes(8, 'little'),
            self.target_root
        ])

def generate_random_bytes(length: int) -> bytes:
    """Generate random bytes of specified length"""
    return bytes(random.getrandbits(8) for _ in range(length))

def generate_keypair() -> Tuple[int, bytes]:
    """Generate a BLS keypair (private key, public key)"""
    private_key = bls.KeyGen(generate_random_bytes(32))
    public_key = bls.SkToPk(private_key)
    return private_key, public_key

def generate_test_attestations(count: int) -> List[Tuple[AttestationData, bytes, bytes]]:
    """
    Generate test attestations with valid signatures
    Returns a list of (attestation_data, signature, public_key) tuples
    """
    attestations = []
    
    for _ in range(count):
        # Generate random attestation data
        attestation = AttestationData(
            slot=random.randint(0, 2**32),
            index=random.randint(0, 2**16),
            beacon_block_root=generate_random_bytes(32),
            source_epoch=random.randint(0, 2**32),
            source_root=generate_random_bytes(32),
            target_epoch=random.randint(0, 2**32),
            target_root=generate_random_bytes(32)
        )
        
        # Generate a keypair and sign the attestation
        private_key, public_key = generate_keypair()
        message = attestation.serialize()
        signature = bls.Sign(private_key, message)
        
        attestations.append((attestation, signature, public_key))
    
    return attestations

def run_verification_benchmark(attestations: List[Tuple[AttestationData, bytes, bytes]], 
                              duration_seconds: int = 5) -> int:
    """
    Run a benchmark verifying BLS signatures for the given duration
    Returns the number of verifications performed
    """
    count = 0
    end_time = time.time() + duration_seconds
    
    # Keep verifying signatures until the time is up
    while time.time() < end_time:
        # Get an attestation from the list (cycling through them)
        attestation, signature, public_key = attestations[count % len(attestations)]
        message = attestation.serialize()
        
        # Verify the signature
        assert bls.Verify(public_key, message, signature)
        
        count += 1
    
    return count

def main():
    print("BLS Signature Verification Benchmark for Ethereum Attestations")
    print("=" * 70)
    
    # Parameters
    num_attestations = 100  # Number of test attestations to generate
    trial_duration = 5      # Duration of each trial in seconds
    num_trials = 3          # Number of trials to run
    
    print(f"Generating {num_attestations} test attestations...")
    attestations = generate_test_attestations(num_attestations)
    print("Test attestations generated successfully.")
    
    # Run benchmark trials
    results = []
    print(f"\nRunning {num_trials} benchmark trials (each {trial_duration} seconds):")
    
    for i in range(num_trials):
        print(f"  Trial {i+1}/{num_trials}... ", end="", flush=True)
        verifications = run_verification_benchmark(attestations, trial_duration)
        verifications_per_second = verifications / trial_duration
        results.append(verifications_per_second)
        print(f"{verifications_per_second:.2f} verifications/second")
    
    # Calculate and display statistics
    avg_verifications = statistics.mean(results)
    median_verifications = statistics.median(results)
    stdev_verifications = statistics.stdev(results) if len(results) > 1 else 0
    
    print("\nResults:")
    print(f"  Average: {avg_verifications:.2f} verifications/second")
    print(f"  Median:  {median_verifications:.2f} verifications/second")
    print(f"  Std Dev: {stdev_verifications:.2f}")
    
    # Individual trial results
    print("\nIndividual trial results (verifications/second):")
    for i, result in enumerate(results):
        print(f"  Trial {i+1}: {result:.2f}")

if __name__ == "__main__":
    main()

File: go/bls_benchmark.go
package main

import (
	"crypto/rand"
	"encoding/binary"
	"fmt"
	"math"
	"sort"
	"time"

	"github.com/herumi/bls-eth-go-binary/bls"
)

// AttestationData represents a simplified Ethereum attestation
type AttestationData struct {
	Slot           uint64
	Index          uint64
	BeaconBlockRoot [32]byte
	SourceEpoch    uint64
	SourceRoot     [32]byte
	TargetEpoch    uint64
	TargetRoot     [32]byte
}

// Serialize converts the attestation data to bytes for signing
func (a *AttestationData) Serialize() []byte {
	buf := make([]byte, 0, 128)

	slotBytes := make([]byte, 8)
	binary.LittleEndian.PutUint64(slotBytes, a.Slot)
	buf = append(buf, slotBytes...)

	indexBytes := make([]byte, 8)
	binary.LittleEndian.PutUint64(indexBytes, a.Index)
	buf = append(buf, indexBytes...)

	buf = append(buf, a.BeaconBlockRoot[:]...)

	sourceEpochBytes := make([]byte, 8)
	binary.LittleEndian.PutUint64(sourceEpochBytes, a.SourceEpoch)
	buf = append(buf, sourceEpochBytes...)

	buf = append(buf, a.SourceRoot[:]...)

	targetEpochBytes := make([]byte, 8)
	binary.LittleEndian.PutUint64(targetEpochBytes, a.TargetEpoch)
	buf = append(buf, targetEpochBytes...)

	buf = append(buf, a.TargetRoot[:]...)

	return buf
}

// SignedAttestation represents an attestation with its signature and public key
type SignedAttestation struct {
	Data      AttestationData
	Signature bls.Sign
	PublicKey bls.PublicKey
}

// generateRandomBytes creates random bytes of specified length
func generateRandomBytes(length int) []byte {
	bytes := make([]byte, length)
	_, err := rand.Read(bytes)
	if err != nil {
		panic(err)
	}
	return bytes
}

// generateRandomAttestation creates a random attestation data
func generateRandomAttestation() AttestationData {
	var blockRoot, sourceRoot, targetRoot [32]byte
	copy(blockRoot[:], generateRandomBytes(32))
	copy(sourceRoot[:], generateRandomBytes(32))
	copy(targetRoot[:], generateRandomBytes(32))

	return AttestationData{
		Slot:           binary.LittleEndian.Uint64(generateRandomBytes(8)),
		Index:          binary.LittleEndian.Uint64(generateRandomBytes(8)) % 65536,
		BeaconBlockRoot: blockRoot,
		SourceEpoch:    binary.LittleEndian.Uint64(generateRandomBytes(8)),
		SourceRoot:     sourceRoot,
		TargetEpoch:    binary.LittleEndian.Uint64(generateRandomBytes(8)),
		TargetRoot:     targetRoot,
	}
}

// generateTestAttestations creates a set of signed attestations for testing
func generateTestAttestations(count int) []SignedAttestation {
	attestations := make([]SignedAttestation, count)

	for i := 0; i < count; i++ {
		// Generate random attestation data
		attestation := generateRandomAttestation()

		// Generate a keypair
		var secretKey bls.SecretKey
		secretKey.SetByCSPRNG()
		publicKey := secretKey.GetPublicKey()

		// Sign the attestation
		message := attestation.Serialize()
		signature := secretKey.Sign(string(message))

		attestations[i] = SignedAttestation{
			Data:      attestation,
			Signature: *signature,
			PublicKey: *publicKey,
		}
	}

	return attestations
}

// runVerificationBenchmark measures how many verifications can be done in the given duration
func runVerificationBenchmark(attestations []SignedAttestation, durationSeconds int) int {
	count := 0
	endTime := time.Now().Add(time.Duration(durationSeconds) * time.Second)

	// Keep verifying signatures until the time is up
	for time.Now().Before(endTime) {
		// Get an attestation from the list (cycling through them)
		attestation := attestations[count%len(attestations)]
		message := attestation.Data.Serialize()

		// Verify the signature
		if !attestation.Signature.Verify(&attestation.PublicKey, string(message)) {
			panic("Signature verification failed")
		}

		count++
	}

	return count
}

// runBatchVerificationBenchmark measures how many batch verifications can be done
// Since the library doesn't have a built-in batch verification, we'll verify each signature
// but time how long it takes to verify a batch of signatures
func runBatchVerificationBenchmark(attestations []SignedAttestation, durationSeconds int, batchSize int) int {
	count := 0
	totalVerifications := 0
	endTime := time.Now().Add(time.Duration(durationSeconds) * time.Second)

	// Keep verifying signatures until the time is up
	for time.Now().Before(endTime) {
		// Process a batch at a time
		batchStart := count % len(attestations)
		allValid := true

		// Verify each signature in the batch
		for i := 0; i < batchSize; i++ {
			idx := (batchStart + i) % len(attestations)
			attestation := attestations[idx]
			message := attestation.Data.Serialize()

			// Verify the signature
			if !attestation.Signature.Verify(&attestation.PublicKey, string(message)) {
				allValid = false
				break
			}
		}

		if !allValid {
			panic("Batch signature verification failed")
		}

		count += batchSize
		totalVerifications += batchSize
	}

	return totalVerifications
}

// calculateStats computes statistics for the benchmark results
func calculateStats(results []float64) (float64, float64, float64) {
	// Calculate mean
	sum := 0.0
	for _, r := range results {
		sum += r
	}
	mean := sum / float64(len(results))

	// Calculate median
	sortedResults := make([]float64, len(results))
	copy(sortedResults, results)
	sort.Float64s(sortedResults)
	
	var median float64
	if len(sortedResults)%2 == 0 {
		median = (sortedResults[len(sortedResults)/2-1] + sortedResults[len(sortedResults)/2]) / 2
	} else {
		median = sortedResults[len(sortedResults)/2]
	}

	// Calculate standard deviation
	sumSquaredDiff := 0.0
	for _, r := range results {
		diff := r - mean
		sumSquaredDiff += diff * diff
	}
	stdDev := math.Sqrt(sumSquaredDiff / float64(len(results)))

	return mean, median, stdDev
}

func main() {
	// Initialize the BLS library
	err := bls.Init(bls.BLS12_381)
	if err != nil {
		panic(err)
	}
	bls.SetETHmode(bls.EthModeDraft07)

	fmt.Println("BLS Signature Verification Benchmark for Ethereum Attestations")
	fmt.Println("======================================================================")

	// Parameters
	numAttestations := 100 // Number of test attestations to generate
	trialDuration := 5     // Duration of each trial in seconds
	numTrials := 3         // Number of trials to run
	batchSizes := []int{1, 10, 50, 100} // Batch sizes to test

	fmt.Printf("Generating %d test attestations...\n", numAttestations)
	attestations := generateTestAttestations(numAttestations)
	fmt.Println("Test attestations generated successfully.")

	// Run individual verification benchmark
	fmt.Printf("\nRunning %d individual verification benchmark trials (each %d seconds):\n", numTrials, trialDuration)
	individualResults := make([]float64, numTrials)

	for i := 0; i < numTrials; i++ {
		fmt.Printf("  Trial %d/%d... ", i+1, numTrials)
		verifications := runVerificationBenchmark(attestations, trialDuration)
		verificationsPerSecond := float64(verifications) / float64(trialDuration)
		individualResults[i] = verificationsPerSecond
		fmt.Printf("%.2f verifications/second\n", verificationsPerSecond)
	}

	// Calculate and display statistics for individual verification
	avgVerifications, medianVerifications, stdDevVerifications := calculateStats(individualResults)

	fmt.Println("\nIndividual Verification Results:")
	fmt.Printf("  Average: %.2f verifications/second\n", avgVerifications)
	fmt.Printf("  Median:  %.2f verifications/second\n", medianVerifications)
	fmt.Printf("  Std Dev: %.2f\n", stdDevVerifications)

	// Run batch verification benchmarks
	fmt.Println("\nBatch Verification Results:")
	
	for _, batchSize := range batchSizes {
		fmt.Printf("\nBatch size: %d\n", batchSize)
		fmt.Printf("Running %d batch verification benchmark trials (each %d seconds):\n", numTrials, trialDuration)
		
		batchResults := make([]float64, numTrials)
		
		for i := 0; i < numTrials; i++ {
			fmt.Printf("  Trial %d/%d... ", i+1, numTrials)
			verifications := runBatchVerificationBenchmark(attestations, trialDuration, batchSize)
			verificationsPerSecond := float64(verifications) / float64(trialDuration)
			batchResults[i] = verificationsPerSecond
			fmt.Printf("%.2f verifications/second\n", verificationsPerSecond)
		}
		
		// Calculate and display statistics for batch verification
		avgBatchVerifications, medianBatchVerifications, stdDevBatchVerifications := calculateStats(batchResults)
		
		fmt.Printf("\n  Batch Size %d Results:\n", batchSize)
		fmt.Printf("    Average: %.2f verifications/second\n", avgBatchVerifications)
		fmt.Printf("    Median:  %.2f verifications/second\n", medianBatchVerifications)
		fmt.Printf("    Std Dev: %.2f\n", stdDevBatchVerifications)
		
		// Calculate and display speedup compared to individual verification
		speedup := avgBatchVerifications / avgVerifications
		fmt.Printf("    Speedup: %.2fx compared to individual verification\n", speedup)
	}

	// Individual trial results
	fmt.Println("\nIndividual verification trial results (verifications/second):")
	for i, result := range individualResults {
		fmt.Printf("  Trial %d: %.2f\n", i+1, result)
	}
}
